File: app/src/main/java/com/hmproductions/theredstreet/fragment/MortgageFragment.java
package com.hmproductions.theredstreet.fragment;


import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioGroup;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.hmproductions.theredstreet.R;
import com.hmproductions.theredstreet.dagger.ContextModule;
import com.hmproductions.theredstreet.dagger.DaggerDalalStreetApplicationComponent;
import com.hmproductions.theredstreet.loaders.MortgageDetailsLoader;
import com.hmproductions.theredstreet.ui.MainActivity;
import com.hmproductions.theredstreet.utils.Constants;
import com.hmproductions.theredstreet.utils.StockUtils;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import dalalstreet.api.DalalActionServiceGrpc;
import dalalstreet.api.actions.GetMortgageDetailsResponse;
import dalalstreet.api.actions.MortgageStocksRequest;
import dalalstreet.api.actions.MortgageStocksResponse;
import dalalstreet.api.actions.RetrieveMortgageStocksRequest;
import dalalstreet.api.actions.RetrieveMortgageStocksResponse;

import static com.hmproductions.theredstreet.utils.StockUtils.getCompanyNameFromStockId;
import static com.hmproductions.theredstreet.utils.StockUtils.getQuantityOwnedFromCompanyName;

/* Uses GetMortgageDetails() for setting stocksMortgaged (int data member)
*  Uses MortgageStocks() to mortgage stocks
*  Uses RetrieveStocksFromMortgage() to get back mortgaged stocks */
public class MortgageFragment extends Fragment implements LoaderManager.LoaderCallbacks<GetMortgageDetailsResponse>{

    @Inject
    DalalActionServiceGrpc.DalalActionServiceBlockingStub actionServiceBlockingStub;

    @BindView(R.id.mortgageRadioGroup)
    RadioGroup mortgageRadioGroup;

    @BindView(R.id.stocks_editText)
    EditText stocksEditText;

    @BindView(R.id.stocksOwned_textView)
    TextView ownedTextView;

    @BindView(R.id.stocksMortgaged_textView)
    TextView mortgagedTextView;

    @BindView(R.id.currentPrice_textView)
    TextView currentPriceTextView;

    @BindView(R.id.mortgageDeposit_textView)
    TextView mortgageDepositTextView;

    @BindView(R.id.mortgageRetrieve_textView)
    TextView mortgageRetrieveTextView;

    int stocksOwned = 0, stocksMortgaged = 0, stocksTransaction;
    Spinner companySpinner;
    String [] companiesArray;
    private AlertDialog loadingDialog;

    public MortgageFragment() {
        // Required empty public constructor
    }

    private BroadcastReceiver refreshStockPricesReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (getActivity() != null && intent.getAction() != null && intent.getAction().equalsIgnoreCase(Constants.REFRESH_STOCK_PRICES_ACTION)) {
                getActivity().getSupportLoaderManager().restartLoader(Constants.MORTGAGE_DETAILS_LOADER_ID, null, MortgageFragment.this);
            }
        }
    };

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (getContext() != null){
            View dialogView = LayoutInflater.from(getContext()).inflate(R.layout.progress_dialog, null);
            String tempString = "Getting mortgage details...";
            ((TextView) dialogView.findViewById(R.id.progressDialog_textView)).setText(tempString);
            loadingDialog = new AlertDialog.Builder(getContext())
                    .setView(dialogView)
                    .setCancelable(false)
                    .create();
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        final View rootView = inflater.inflate(R.layout.fragment_mortgage, container, false);

        DaggerDalalStreetApplicationComponent.builder().contextModule(new ContextModule(getContext())).build().inject(this);
        ButterKnife.bind(this, rootView);
        if (getActivity() != null) getActivity().setTitle("Mortgage Stocks");

        companiesArray = StockUtils.getCompanyNamesArray();

        companySpinner = rootView.findViewById(R.id.mortgage_companies_spinner);
        ArrayAdapter<String> arrayAdapter = new ArrayAdapter<>(getActivity(), R.layout.company_spinner_item, companiesArray);
        companySpinner.setAdapter(arrayAdapter);

        companySpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                getActivity().getSupportLoaderManager().restartLoader(Constants.MORTGAGE_DETAILS_LOADER_ID, null, MortgageFragment.this);
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });

        mortgageRadioGroup.setOnCheckedChangeListener((radioGroup, id) -> {
            Button mortgageButton = rootView.findViewById(R.id.mortgage_button);
            mortgageButton.setText(id==R.id.mortgage_radioButton?"MORTGAGE":"RETRIEVE");
        });

        String ownedString = "N/A";
        ownedTextView.setText(ownedString);

        String mortgageString = "N/A";
        mortgagedTextView.setText(mortgageString);

        return rootView;
    }

    @OnClick(R.id.mortgage_button)
    void onMortgageButtonClick() {

        if (stocksEditText.getText().toString().trim().isEmpty()) {
            stocksEditText.setError("Stocks quantity missing");
            stocksEditText.requestFocus();
            return;
        } else {
            stocksEditText.setError(null);
            stocksEditText.clearFocus();
        }

        stocksTransaction = Integer.parseInt(stocksEditText.getText().toString().trim());

        if (mortgageRadioGroup.getCheckedRadioButtonId() == R.id.mortgage_radioButton) {

            if (stocksTransaction <= stocksOwned) {
                MortgageStocksResponse mortgageStocksResponse = actionServiceBlockingStub.mortgageStocks(
                        MortgageStocksRequest
                                .newBuilder()
                                .setStockId(companySpinner.getSelectedItemPosition()+1)
                                .setStockQuantity(Integer.parseInt(stocksEditText.getText().toString()))
                                .build()
                );

                if (mortgageStocksResponse.getStatusCode().getNumber() == 0) {
                    Toast.makeText(getContext(), "Transaction successful", Toast.LENGTH_SHORT).show();

                    stocksOwned -= stocksTransaction;
                    stocksMortgaged += stocksTransaction;

                    String ownedString = " :  " + String.valueOf(stocksOwned);
                    ownedTextView.setText(ownedString);

                    String mortgageString = " :  " + String.valueOf(stocksMortgaged);
                    mortgagedTextView.setText(mortgageString);

                    stocksEditText.setText("");

                } else if (mortgageStocksResponse.getStatusCode().getNumber() == 2) {
                    Toast.makeText(getContext(), "Market is Closed", Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(getContext(), "Inconsistent data server error", Toast.LENGTH_SHORT).show();
                }

            } else {
                Toast.makeText(getActivity(), "Insufficient Stocks", Toast.LENGTH_SHORT).show();
            }

        } else if (mortgageRadioGroup.getCheckedRadioButtonId() == R.id.retrieve_radioButton) {

            if (stocksTransaction <= stocksMortgaged && stocksMortgaged >= 0) {
                RetrieveMortgageStocksResponse retrieveStocksResponse = actionServiceBlockingStub.retrieveMortgageStocks(
                        RetrieveMortgageStocksRequest
                                .newBuilder()
                                .setStockId(companySpinner.getSelectedItemPosition()+1)
                                .setStockQuantity(Integer.parseInt(stocksEditText.getText().toString()))
                                .build()
                );

                if (retrieveStocksResponse.getStatusCode().getNumber() == 0) {
                    Toast.makeText(getContext(), "Transaction successful", Toast.LENGTH_SHORT).show();

                    stocksOwned += stocksTransaction;
                    stocksMortgaged -= stocksTransaction;

                    String ownedString = " :  " + String.valueOf(stocksOwned);
                    ownedTextView.setText(ownedString);

                    String mortgageString = " :  " + String.valueOf(stocksMortgaged);
                    mortgagedTextView.setText(mortgageString);

                    stocksEditText.setText("");

                } else if (retrieveStocksResponse.getStatusCode().getNumber() == 2) {
                    Toast.makeText(getContext(), "Market is Closed", Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(getContext(), "Inconsistent data server error", Toast.LENGTH_SHORT).show();
                }

            } else {
                Toast.makeText(getActivity(), "You dont have sufficient stocks mortgaged", Toast.LENGTH_SHORT).show();
            }
        } else {
            Toast.makeText(getContext(), "Select mortgage or retrieve", Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public Loader<GetMortgageDetailsResponse> onCreateLoader(int id, Bundle args) {
        loadingDialog.show();
        companySpinner.setEnabled(false);

        if (getContext() != null) {
            return new MortgageDetailsLoader(getContext(), actionServiceBlockingStub);
        } else {
            return null;
        }
    }

    @Override
    public void onLoadFinished(Loader<GetMortgageDetailsResponse> loader, GetMortgageDetailsResponse response) {

        loadingDialog.dismiss();

        int stockId = StockUtils.getStockIdFromCompanyName(companiesArray[companySpinner.getSelectedItemPosition()]);
        companySpinner.setEnabled(true);

        if (response.getMortgageMapMap().get(stockId) != null)
            stocksMortgaged = response.getMortgageMapMap().get(stockId);
        else
            stocksMortgaged = 0;

        String mortgageString = " :  " + String.valueOf(stocksMortgaged);
        mortgagedTextView.setText(mortgageString);

        stocksOwned = getQuantityOwnedFromCompanyName(MainActivity.ownedStockDetails, getCompanyNameFromStockId(stockId));

        String ownedString = " :  " + String.valueOf(stocksOwned);
        ownedTextView.setText(ownedString);

        int currentPrice = StockUtils.getPriceFromStockId(MainActivity.globalStockDetails, stockId);
        String tempString = " :  " + Constants.RUPEE_SYMBOL + " " + String.valueOf(currentPrice);
        currentPriceTextView.setText(tempString);

        tempString = " :  " + Constants.RUPEE_SYMBOL + " " + String.valueOf(Constants.MORTGAGE_DEPOSIT_RATE * currentPrice / 100);
        mortgageDepositTextView.setText(tempString);

        tempString = " :  " + Constants.RUPEE_SYMBOL + " " + String.valueOf(Constants.MORTGAGE_RETRIEVE_RATE * currentPrice / 100);
        mortgageRetrieveTextView.setText(tempString);
    }

    @Override
    public void onLoaderReset(Loader<GetMortgageDetailsResponse> loader) {
        // Do nothing
    }

    @Override
    public void onResume() {
        super.onResume();
        if (getContext() != null) {
            LocalBroadcastManager.getInstance(getContext()).registerReceiver(
                    refreshStockPricesReceiver, new IntentFilter(Constants.REFRESH_STOCK_PRICES_ACTION)
            );
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        if (getContext() != null) {
            LocalBroadcastManager.getInstance(getContext()).unregisterReceiver(refreshStockPricesReceiver);
        }
    }
}

File: app/src/main/java/org/pragyan/dalal18/fragment/mortgage/MortgageFragment.kt
package org.pragyan.dalal18.fragment.mortgage

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import dalalstreet.api.DalalActionServiceGrpc
import dalalstreet.api.actions.MortgageStocksRequest
import dalalstreet.api.actions.MortgageStocksResponse
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.withContext
import org.pragyan.dalal18.R
import org.pragyan.dalal18.dagger.ContextModule
import org.pragyan.dalal18.dagger.DaggerDalalStreetApplicationComponent
import org.pragyan.dalal18.data.DalalViewModel
import org.pragyan.dalal18.databinding.FragmentMortgageBinding
import org.pragyan.dalal18.ui.MainActivity.Companion.GAME_STATE_UPDATE_ACTION
import org.pragyan.dalal18.utils.*
import java.text.DecimalFormat
import javax.inject.Inject

class MortgageFragment : Fragment() {

    private var binding by viewLifecycle<FragmentMortgageBinding>()

    @Inject
    lateinit var actionServiceBlockingStub: DalalActionServiceGrpc.DalalActionServiceBlockingStub

    private var lastStockId = 1
    private val decimalFormat = DecimalFormat(Constants.PRICE_FORMAT)
    private lateinit var model: DalalViewModel

    lateinit var networkDownHandler: ConnectionUtils.OnNetworkDownHandler
    private var loadingDialog: AlertDialog? = null

    private val refreshStockPricesReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {

            if (intent.action.equals(Constants.REFRESH_STOCK_PRICES_FOR_ALL, ignoreCase = true) ||
                    intent.action.equals(Constants.REFRESH_STOCKS_FOR_MORTGAGE, ignoreCase = true) ||
                    intent.action.equals(GAME_STATE_UPDATE_ACTION, ignoreCase = true)) {
                setupMortgageDetails(lastStockId)
            }
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            networkDownHandler = context as ConnectionUtils.OnNetworkDownHandler
        } catch (classCastException: ClassCastException) {
            throw ClassCastException("$context must implement network down handler.")
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentMortgageBinding.inflate(inflater, container, false)

        model = activity?.run { ViewModelProvider(this).get(DalalViewModel::class.java) }
                ?: throw Exception("Invalid activity")
        lastStockId = model.favoriteCompanyStockId ?: 1
        DaggerDalalStreetApplicationComponent.builder().contextModule(ContextModule(context!!)).build().inject(this)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        (activity as AppCompatActivity).supportActionBar?.title = resources.getString(R.string.mortgage)

        val dialogView = LayoutInflater.from(context).inflate(R.layout.progress_dialog, null)
        val tempString = "Getting mortgage details..."
        (dialogView.findViewById<View>(R.id.progressDialog_textView) as TextView).text = tempString
        loadingDialog = AlertDialog.Builder(context!!).setView(dialogView).setCancelable(false).create()

        setupMortgageDetails(lastStockId)

        binding.apply {
            with(mortgageCompaniesSpinner) {
                val companiesArray = model.getCompanyNamesArray()
                adapter = ArrayAdapter(context!!, R.layout.company_spinner_item, companiesArray)

                onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                    override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {

                        lastStockId = model.getStockIdFromCompanyName(companiesArray[position])

                        model.updateFavouriteCompanyStockId(lastStockId)

                        setupMortgageDetails(lastStockId)
                    }

                    override fun onNothingSelected(parent: AdapterView<*>) {

                    }
                }
            }

            mortgageButton.setOnClickListener { onMortgageButtonClick() }
            stockIncrementFiveButton.setOnClickListener { addToMortgageInput(5) }
            stockIncrementOneButton.setOnClickListener { addToMortgageInput(1) }

            val tempText = " :  ${Constants.MORTGAGE_DEPOSIT_RATE}%"
            depositRateTextView.text = tempText
        }
    }

    private fun addToMortgageInput(increment: Int) {
        binding.apply {
            if (mortgageStocksEditText.text.isBlank())
                mortgageStocksEditText.setText("0")
            var noOfStocks = mortgageStocksEditText.text.toString().toInt()
            noOfStocks = (noOfStocks + increment).coerceAtMost(Constants.ASK_LIMIT)
            mortgageStocksEditText.setText(noOfStocks.toString())
        }
    }

    private fun onMortgageButtonClick() {
        with(binding.mortgageStocksEditText) {
            when {
                text.toString().trim { it <= ' ' }.isEmpty() -> {
                    error = "Stocks quantity missing"
                    requestFocus()
                    return
                }
                text.toString().toLong() == 0L -> {
                    error = "Enter valid number of stocks"
                    requestFocus()
                    return
                }
                else -> {
                    error = null
                    clearFocus()
                }
            }
        }

        val stocksTransaction = (binding.mortgageStocksEditText.text.toString().trim { it <= ' ' }).toLong()

        if (stocksTransaction <= model.getQuantityOwnedFromStockId(lastStockId))
            mortgageStocksAsynchronously(stocksTransaction)
        else context?.toast("Insufficient Stocks")
    }

    private fun setupMortgageDetails(stockId: Int) {
        binding.apply {
            val ownedString = " :  " + decimalFormat.format(model.getQuantityOwnedFromStockId(stockId))
            stocksOwnedTextView.text = ownedString

            val tempString = " :  " + Constants.RUPEE_SYMBOL + " " + decimalFormat.format(model.getPriceFromStockId(stockId))
            currentPriceTextView.text = tempString

            val mortgagedString = " :  " + decimalFormat.format(model.getMortgagedStocksFromStockId(stockId))
            stocksMortgagedTextView.text = mortgagedString

            when {
                model.getIsBankruptFromStockId(lastStockId) -> companyStatusIndicatorImageView.setStatusIndicator(context, View.VISIBLE, getString(R.string.this_company_is_bankrupt), R.drawable.bankrupt_icon)
                model.getGivesDividendFromStockId(lastStockId) -> companyStatusIndicatorImageView.setStatusIndicator(context, View.VISIBLE, getString(R.string.this_company_gives_dividend), R.drawable.dividend_icon)
                else -> companyStatusIndicatorImageView.setStatusIndicator(context, View.GONE, "", R.drawable.clear_icon)
            }
        }
    }

    private fun mortgageStocksAsynchronously(stocksTransaction: Long) {
        val dialogView = LayoutInflater.from(context).inflate(R.layout.progress_dialog, null)
        val tempString = "Mortgaging Stocks..."
        (dialogView.findViewById<View>(R.id.progressDialog_textView) as TextView).text = tempString
        loadingDialog = AlertDialog.Builder(context!!).setView(dialogView).setCancelable(false).create()
        loadingDialog?.show()
        GlobalScope.async (Dispatchers.Default){

            if (ConnectionUtils.getConnectionInfo(context)) {
                if (ConnectionUtils.isReachableByTcp(Constants.HOST, Constants.PORT)) {
                    val mortgageStocksResponse = actionServiceBlockingStub.mortgageStocks(
                            MortgageStocksRequest.newBuilder().setStockId(lastStockId)
                                    .setStockQuantity(stocksTransaction).build())

                    withContext(Dispatchers.Main) {
                        if (mortgageStocksResponse.statusCode == MortgageStocksResponse.StatusCode.OK) {
                            context?.toast("Transaction successful")
                            binding.mortgageStocksEditText.setText("0")
                            view?.hideKeyboard()
                        } else {
                            context?.toast(mortgageStocksResponse.statusMessage)
                        }
                    }
                } else {
                    withContext(Dispatchers.Main) { networkDownHandler.onNetworkDownError(resources.getString(R.string.error_server_down), R.id.main_mortgage_dest) }
                }
            } else {
                withContext(Dispatchers.Main) { networkDownHandler.onNetworkDownError(resources.getString(R.string.error_check_internet), R.id.main_mortgage_dest) }
            }
            withContext(Dispatchers.Main) { loadingDialog?.dismiss() }
        }
    }

    override fun onResume() {
        super.onResume()

        binding.mortgageCompaniesSpinner.setSelection(model.getIndexForFavoriteCompany())

        val intentFilter = IntentFilter(Constants.REFRESH_STOCK_PRICES_FOR_ALL)
        intentFilter.addAction(Constants.REFRESH_STOCKS_FOR_MORTGAGE)
        intentFilter.addAction(GAME_STATE_UPDATE_ACTION)
        LocalBroadcastManager.getInstance(context!!).registerReceiver(refreshStockPricesReceiver, intentFilter)
    }

    override fun onPause() {
        super.onPause()
        LocalBroadcastManager.getInstance(context!!).unregisterReceiver(refreshStockPricesReceiver)
    }
}
